dispatch/httpdispatch.go

```diff
@@ -91,65 +77,109 @@ func NewDispatcher(dc DispatcherConfig, norn model.Norn, transport http.RoundTri
-	}
+func (h HttpDispatcher) Start(_ context.Context) error {
+	return nil
 
-	if config.SecretKey == "" {
-		return fmt.Errorf("invalid AWS secretkey")
-	}
+}
 
-	if config.ID == "" {
-		return fmt.Errorf("invalid AWS id")
+// called to deliver event
+func (h HttpDispatcher) Send(msg *wrp.Message) {
+	defer func() {
+		h.Wg.Done()
+		if r := recover(); nil != r {
+			h.Measures.DroppedPanicCounter.Add(1.0)
+			h.Logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), "goroutine send() panicked",
+				"id", h.Norn.DeviceID, "panic", r)
+		}
+		h.Workers.Release()
+		h.Measures.WorkersCount.Add(-1.0)
+	}()
+
+	var (
+		url   string
+		code  string
+		event string
+	)
+
+	buffer := bytes.NewBuffer([]byte{})
+	encoder := wrp.NewEncoder(buffer, wrp.JSON)
+	err := encoder.Encode(msg)
+	if err != nil {
+		h.Measures.DroppedInvalidConfig.Add(1.0)
+		h.Logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), "Failed to marshal event.")
+		return
+	}
+	bodyPayload := buffer.Bytes()
+	payloadReader := bytes.NewReader(bodyPayload)
+
+	var (
+		body []byte
+	)
+	req, err := http.NewRequest("POST", h.Norn.Destination.HttpConfig.URL, payloadReader)
+	if nil != err {
+		h.Measures.DroppedInvalidConfig.Add(1.0)
+		h.Logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), "Invalid URL",
+			"url", h.Norn.Destination.HttpConfig.URL, logging.ErrorKey(), err)
+		return
 	}
 
-	if config.Sqs.QueueURL == "" {
-		return fmt.Errorf("invalid SQS queueUrl")
+	// Apply the secret
+	if "" != h.Norn.Destination.HttpConfig.Secret {
+		s := hmac.New(sha1.New, []byte(h.Norn.Destination.HttpConfig.Secret))
+		s.Write(body)
+		sig := fmt.Sprintf("sha1=%s", hex.EncodeToString(s.Sum(nil)))
+		req.Header.Set("X-Codex-Signature", sig)
 	}
 
-	if config.Sqs.Region == "" {
+	// find the event "short name"
+	event = msg.FindEventStringSubMatch()
 
+	retryOptions := xhttp.RetryOptions{
+		Logger:      h.Logger,
+		Retries:     h.DeliveryRetries,
+		Interval:    h.DeliveryInterval,
+		ShouldRetry: func(error) bool { return true },
+		ShouldRetryStatus: func(code int) bool {
+			return code < 200 || code > 299
+		},
 	}
+	resp, err := xhttp.RetryTransactor(retryOptions, h.Sender)(req)
+	code = "failure"
+	if nil != err {
+		h.Measures.DroppedNetworkErrCounter.Add(1.0)
+	} else {
+		// Report Result for metrics
+		code = strconv.Itoa(resp.StatusCode)
+
+		// read until the response is complete before closing to allow
+		// connection reuse
+		if nil != resp.Body {
+			io.Copy(ioutil.Discard, resp.Body)
+			resp.Body.Close()
+		}
+	}
+	url = h.Norn.Destination.HttpConfig.URL
+	h.Measures.DeliveryCounter.With("url", url, "code", code, "event", event).Add(1.0)
 
-	return nil
 }
 
-func NewEventWithID(event *wrp.Message, deviceID string) *eventWithID {
-	return &eventWithID{
-		Event:    event,
-		DeviceID: deviceID,
+func (h HttpDispatcher) Update(norn model.Norn) {
+
+	if h.Norn.Destination.HttpConfig.Secret != norn.Destination.HttpConfig.Secret {
```

Comment:
The config should be protected by the read write mutex; so anytime we reference or change it, it should be surrounded by a lock() and unlock() of some sort.